manul ignores build tags?

This is on mac.

You can repro with this code: https://github.com/kjk/web-blog (`git clone --recursive https://github.com/kjk/web-blog`)

```
$ manul -U -r
2016/10/22 14:30:37 updating dependency github.com/garyburd/go-oauth
2016/10/22 14:30:37 updating dependency github.com/go-fsnotify/fsnotify
2016/10/22 14:30:38 updating dependency github.com/gorilla/securecookie
2016/10/22 14:30:38 updating dependency github.com/gorilla/websocket
2016/10/22 14:30:39 updating dependency github.com/kjk/u
2016/10/22 14:30:39 updating dependency github.com/kr/fs
2016/10/22 14:30:40 updating dependency github.com/microcosm-cc/bluemonday
2016/10/22 14:30:40 updating dependency github.com/russross/blackfriday
2016/10/22 14:30:41 updating dependency github.com/shurcooL/sanitized_anchor_name
2016/10/22 14:30:42 updating dependency github.com/thomas11/atomgenerator
2016/10/22 14:30:42 updating dependency golang.org/x/net
updated 11 dependencies
$ go build
vendor/github.com/go-fsnotify/fsnotify/kqueue.go:18:2: cannot find package "golang.org/x/sys/unix" in any of:
    /Users/kjk/src/go/src/github.com/kjk/blog/vendor/golang.org/x/sys/unix (vendor tree)
    /usr/local/Cellar/go/1.7.3/libexec/src/golang.org/x/sys/unix (from $GOROOT)
    /Users/kjk/src/go/src/golang.org/x/sys/unix (from $GOPATH)
kjkmacpro:blog kjk$
```

The problem seems to be that in `kqueue.go` we have:

```
// +build freebsd openbsd netbsd dragonfly darwin

package fsnotify

import (
    "errors"
    "fmt"
    "io/ioutil"
    "os"
    "path/filepath"
    "sync"
    "time"

    "golang.org/x/sys/unix"
)
```

I assume it's the build tag. Since I'm building on mac and build tag specifies darwin, the file should be parsed and `golang.org/x/sys/unix` be vendored as well.

--- a/imports.go
+++ b/imports.go
@@ -1,7 +1,6 @@
 package main
 
 import (
-	"bufio"
 	"encoding/json"
 	"fmt"
 	"os"
@@ -43,13 +42,15 @@ func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 	return imports, nil
 }
 
-func calculateDependencies(packages []string, recursive, testDependencies bool) ([]string, error) {
+func calculateDependencies(packages []string, recursive,
+	testDependencies bool) ([]string, error) {
 	var deps, imports, testImports, testDeps []string
 
 	for _, pkg := range packages {
 		data, err := list(pkg)
 		if err != nil {
-			return imports, fmt.Errorf("failed to list dependencies for %s: %s", pkg, err)
+			return imports,
+				fmt.Errorf("failed to list dependencies for %s: %s", pkg, err)
 		}
 
 		if recursive {
@@ -68,7 +69,9 @@ func calculateDependencies(packages []string, recursive, testDependencies bool) 
 		for _, testImport := range testImports {
 			testData, err := list(testImport)
 			if err != nil {
-				return imports, fmt.Errorf("failed to list dependencies for testing package %s: %s", testImport, err)
+				return imports, fmt.Errorf(
+					"failed to list dependencies for testing package %s: %s",
+					testImport, err)
 			}
 			testDeps = append(testDeps, testData.Deps...)
 		}
@@ -92,7 +95,8 @@ func filterPackages(packages []string) []string {
 		}
 
 		if inTests {
-			importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
+			importpath = strings.Replace(importpath,
+				"__blankd__", "localhost:60001", -1)
 		}
 
 		if isOwnPackage(importpath) {
@@ -122,16 +126,16 @@ func filterPackages(packages []string) []string {
 	return imports
 }
 
-func ensureDependenciesExist(packages []string, includeTestingDependencies bool) error {
+func ensureDependenciesExist(packages []string, includeTestDeps bool) error {
 	args := []string{"get"}
-	if includeTestingDependencies {
+	if includeTestDeps {
 		args = append(args, "-t")
 	}
 	args = append(args, packages...)
 
 	out, err := execute(exec.Command("go", args...))
 	if err != nil {
-		return fmt.Errorf("couldn't 'go get' dependencies:\n%s", out)
+		return fmt.Errorf("can't 'go get' dependencies:\n%s", out)
 	}
 
 	return nil
@@ -147,7 +151,8 @@ func list(pkg string) (listOutput, error) {
 
 	err = json.Unmarshal([]byte(out), &data)
 	if err != nil {
-		return data, fmt.Errorf("can't unmarshal go list JSON output: %s\n%q", err, out)
+		return data,
+			fmt.Errorf("can't unmarshal go list JSON output: %s\n%q", err, out)
 	}
 
 	return data, nil
@@ -161,10 +166,7 @@ func listPackages() ([]string, error) {
 		return packages, err
 	}
 
-	r := strings.NewReader(out)
-	s := bufio.NewScanner(r)
-	for s.Scan() {
-		pkg := s.Text()
+	for _, pkg := range strings.Split(strings.TrimSpace(out), "\n") {
 		if strings.Contains(pkg, "/vendor/") {
 			continue
 		}

--- a/imports.go
+++ b/imports.go
@@ -25,10 +25,10 @@ func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 	}
 
 	// Ensuring our dependencies exists isn't a strict requirement, therefore
-	// only print a message to stderr rather then completely failing
+	// only print a message to stderr rather then completely failing.
 	err = ensureDependenciesExist(packages, testDependencies)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "%s", err)
+		fmt.Fprintf(os.Stderr, "%s\n", err)
 	}
 
 	imports, err = calculateDependencies(packages, recursive, testDependencies)
@@ -84,7 +84,6 @@ func calculateDependencies(packages []string, recursive, testDependencies bool) 
 
 func filterPackages(packages []string) []string {
 	var imports []string
-	cwd, _ := os.Getwd()
 
 	for _, importing := range packages {
 		importpath, err := getRootImportpath(importing)
@@ -96,7 +95,7 @@ func filterPackages(packages []string) []string {
 			importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
 		}
 
-		if isOwnPackage(importpath, cwd) {
+		if isOwnPackage(importpath) {
 			continue
 		}
 
@@ -132,7 +131,7 @@ func ensureDependenciesExist(packages []string, includeTestingDependencies bool)
 
 	out, err := execute(exec.Command("go", args...))
 	if err != nil {
-		return fmt.Errorf("couldn't go get dependencies:\n %s", out)
+		return fmt.Errorf("couldn't 'go get' dependencies:\n%s", out)
 	}
 
 	return nil
@@ -176,9 +175,9 @@ func listPackages() ([]string, error) {
 	return packages, nil
 }
 
-func isOwnPackage(path, cwd string) bool {
+func isOwnPackage(path string) bool {
 	for _, gopath := range filepath.SplitList(os.Getenv("GOPATH")) {
-		if strings.HasPrefix(filepath.Join(gopath, "src", path), cwd) {
+		if strings.HasPrefix(filepath.Join(gopath, "src", path), workdir) {
 			return true
 		}
 	}

--- a/imports.go
+++ b/imports.go
@@ -15,7 +15,7 @@ func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 	var imports []string
 	packages, err := listPackages()
 	if err != nil {
-		return imports, nil
+		return imports, fmt.Errorf("error listing packages: %s", err)
 	}
 
 	ensureDependenciesExist(packages, testDependencies)
@@ -145,7 +145,7 @@ func list(pkg string) (listOutput, error) {
 func listPackages() ([]string, error) {
 	var packages []string
 
-	out, err := execute(exec.Command("go", "list", "./..."))
+	out, err := execute(exec.Command("go", "list", "-e", "./..."))
 	if err != nil {
 		return packages, err
 	}

--- a/imports.go
+++ b/imports.go
@@ -11,6 +11,12 @@ import (
 	"strings"
 )
 
+type listOutput struct {
+	Imports     []string
+	Deps        []string
+	TestImports []string
+}
+
 func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 	var imports []string
 	packages, err := listPackages()
@@ -18,7 +24,12 @@ func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 		return imports, fmt.Errorf("error listing packages: %s", err)
 	}
 
-	ensureDependenciesExist(packages, testDependencies)
+	// Ensuring our dependencies exists isn't a strict requirement, therefore
+	// only print a message to stderr rather then completely failing
+	err = ensureDependenciesExist(packages, testDependencies)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s", err)
+	}
 
 	imports, err = calculateDependencies(packages, recursive, testDependencies)
 	if err != nil {
@@ -30,7 +41,6 @@ func parseImports(recursive bool, testDependencies bool) ([]string, error) {
 	sort.Strings(imports)
 
 	return imports, nil
-
 }
 
 func calculateDependencies(packages []string, recursive, testDependencies bool) ([]string, error) {
@@ -39,7 +49,7 @@ func calculateDependencies(packages []string, recursive, testDependencies bool) 
 	for _, pkg := range packages {
 		data, err := list(pkg)
 		if err != nil {
-			return imports, nil
+			return imports, fmt.Errorf("failed to list dependencies for %s: %s", pkg, err)
 		}
 
 		if recursive {
@@ -55,10 +65,10 @@ func calculateDependencies(packages []string, recursive, testDependencies bool) 
 
 	testImports = unique(testImports)
 	if recursive {
-		for _, i := range testImports {
-			testData, err := list(i)
+		for _, testImport := range testImports {
+			testData, err := list(testImport)
 			if err != nil {
-				return imports, nil
+				return imports, fmt.Errorf("failed to list dependencies for testing package %s: %s", testImport, err)
 			}
 			testDeps = append(testDeps, testData.Deps...)
 		}
@@ -113,7 +123,7 @@ func filterPackages(packages []string) []string {
 	return imports
 }
 
-func ensureDependenciesExist(packages []string, includeTestingDependencies bool) {
+func ensureDependenciesExist(packages []string, includeTestingDependencies bool) error {
 	args := []string{"get"}
 	if includeTestingDependencies {
 		args = append(args, "-t")
@@ -122,8 +132,10 @@ func ensureDependenciesExist(packages []string, includeTestingDependencies bool)
 
 	out, err := execute(exec.Command("go", args...))
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "%s\n", out)
+		return fmt.Errorf("couldn't go get dependencies:\n %s", out)
 	}
+
+	return nil
 }
 
 func list(pkg string) (listOutput, error) {
@@ -136,7 +148,7 @@ func list(pkg string) (listOutput, error) {
 
 	err = json.Unmarshal([]byte(out), &data)
 	if err != nil {
-		return data, err
+		return data, fmt.Errorf("can't unmarshal go list JSON output: %s\n%q", err, out)
 	}
 
 	return data, nil
@@ -174,21 +186,15 @@ func isOwnPackage(path, cwd string) bool {
 }
 
 func unique(input []string) []string {
-	u := make([]string, 0, len(input))
-	m := make(map[string]bool)
+	uniqueList := make([]string, 0, len(input))
+	uniqueMap := make(map[string]bool)
 
 	for _, val := range input {
-		if _, ok := m[val]; !ok {
-			m[val] = true
-			u = append(u, val)
+		if _, ok := uniqueMap[val]; !ok {
+			uniqueMap[val] = true
+			uniqueList = append(uniqueList, val)
 		}
 	}
 
-	return u
-}
-
-type listOutput struct {
-	Imports     []string
-	Deps        []string
-	TestImports []string
+	return uniqueList
 }

--- a/imports.go
+++ b/imports.go
@@ -1,135 +1,167 @@
 package main
 
 import (
-	"go/build"
-	"log"
+	"bufio"
+	"encoding/json"
+	"fmt"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"sort"
 	"strings"
 )
 
-func recursiveParseImports(
-	imports map[string]bool, path string, cwd string,
-) error {
-	if path == "C" {
-		return nil
-	}
-
-	// catch internal vendoring in net/http since go 1.7
-	if strings.HasPrefix(path, "golang_org/") {
-		return nil
-	}
-
-	pkg, err := build.Import(path, cwd, build.IgnoreVendor)
+func parseImports(recursive bool, testDependencies bool) ([]string, error) {
+	var imports []string
+	packages, err := listPackages()
 	if err != nil {
-		return err
+		return imports, nil
 	}
 
-	if path != "." {
-		standard := false
+	ensureDependenciesExist(packages, testDependencies)
 
-		if pkg.Goroot && pkg.ImportPath != "" {
-			standard = true
-		}
-
-		imports[pkg.ImportPath] = standard
-	}
-
-	for _, importing := range pkg.Imports {
-		_, ok := imports[importing]
-		if !ok {
-			err = recursiveParseImports(imports, importing, cwd)
-			if err != nil {
-				return err
-			}
-		}
-	}
-
-	return nil
-}
-
-func parseImports(recursive bool) ([]string, error) {
-	cwd, err := os.Getwd()
+	imports, err = calculateDependencies(packages, recursive, testDependencies)
 	if err != nil {
-		return nil, err
+		return imports, err
 	}
 
-	var (
-		allImports = map[string]bool{}
-		imports    = []string{}
-	)
-
-	filepath.Walk(
-		cwd, func(path string, info os.FileInfo, err error) error {
-			if err != nil {
-				return err
-			}
-
-			if filepath.Base(path) == ".git" ||
-				filepath.Dir(path) == filepath.Join(cwd, "vendor") {
-				return filepath.SkipDir
-			}
-
-			if path == filepath.Join(cwd, "vendor") {
-				return nil
-			}
-
-			if !info.IsDir() {
-				return nil
-			}
-
-			err = recursiveParseImports(
-				allImports,
-				".",
-				path,
-			)
-			if _, ok := err.(*build.NoGoError); ok {
-				return nil
-			}
-
-			if err != nil {
-				log.Println(err)
-			}
-
-			return nil
-		},
-	)
-
-	for importing, standard := range allImports {
-		if !standard {
-			importpath, err := getRootImportpath(importing)
-			if err != nil {
-				continue
-			}
-
-			if inTests {
-				importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
-			}
-
-			if isOwnPackage(importpath, cwd) {
-				continue
-			}
-
-			found := false
-			for _, imported := range imports {
-				if importpath == imported {
-					found = true
-					break
-				}
-			}
-
-			if found {
-				continue
-			}
-
-			imports = append(imports, importpath)
-		}
-	}
+	imports = filterPackages(imports)
 
 	sort.Strings(imports)
 
 	return imports, nil
+
+}
+
+func calculateDependencies(packages []string, recursive, testDependencies bool) ([]string, error) {
+	var deps, imports, testImports, testDeps []string
+
+	for _, pkg := range packages {
+		data, err := list(pkg)
+		if err != nil {
+			return imports, nil
+		}
+
+		if recursive {
+			deps = append(deps, data.Deps...)
+		} else {
+			deps = append(deps, data.Imports...)
+		}
+
+		if testDependencies {
+			testImports = append(testImports, data.TestImports...)
+		}
+	}
+
+	testImports = unique(testImports)
+	if recursive {
+		for _, i := range testImports {
+			testData, err := list(i)
+			if err != nil {
+				return imports, nil
+			}
+			testDeps = append(testDeps, testData.Deps...)
+		}
+
+		deps = append(deps, testDeps...)
+	}
+
+	deps = append(deps, testImports...)
+	deps = unique(deps)
+
+	return deps, nil
+}
+
+func filterPackages(packages []string) []string {
+	var imports []string
+	cwd, _ := os.Getwd()
+
+	for _, importing := range packages {
+		importpath, err := getRootImportpath(importing)
+		if err != nil {
+			continue
+		}
+
+		if inTests {
+			importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
+		}
+
+		if isOwnPackage(importpath, cwd) {
+			continue
+		}
+
+		// Skip anything that is already vendored upstream
+		if strings.Contains(importpath, "/vendor/") {
+			continue
+		}
+
+		found := false
+		for _, imported := range imports {
+			if importpath == imported {
+				found = true
+				break
+			}
+		}
+
+		if found {
+			continue
+		}
+
+		imports = append(imports, importpath)
+	}
+
+	return imports
+}
+
+func ensureDependenciesExist(packages []string, includeTestingDependencies bool) {
+	args := []string{"get"}
+	if includeTestingDependencies {
+		args = append(args, "-t")
+	}
+	args = append(args, packages...)
+
+	out, err := execute(exec.Command("go", args...))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", out)
+	}
+}
+
+func list(pkg string) (listOutput, error) {
+	data := listOutput{}
+
+	out, err := execute(exec.Command("go", "list", "-e", "-json", pkg))
+	if err != nil {
+		return data, err
+	}
+
+	err = json.Unmarshal([]byte(out), &data)
+	if err != nil {
+		return data, err
+	}
+
+	return data, nil
+}
+
+func listPackages() ([]string, error) {
+	var packages []string
+
+	out, err := execute(exec.Command("go", "list", "./..."))
+	if err != nil {
+		return packages, err
+	}
+
+	r := strings.NewReader(out)
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		pkg := s.Text()
+		if strings.Contains(pkg, "/vendor/") {
+			continue
+		}
+
+		packages = append(packages, pkg)
+	}
+
+	return packages, nil
 }
 
 func isOwnPackage(path, cwd string) bool {
@@ -139,4 +171,24 @@ func isOwnPackage(path, cwd string) bool {
 		}
 	}
 	return false
+}
+
+func unique(input []string) []string {
+	u := make([]string, 0, len(input))
+	m := make(map[string]bool)
+
+	for _, val := range input {
+		if _, ok := m[val]; !ok {
+			m[val] = true
+			u = append(u, val)
+		}
+	}
+
+	return u
+}
+
+type listOutput struct {
+	Imports     []string
+	Deps        []string
+	TestImports []string
 }

--- a/main.go
+++ b/main.go
@@ -73,7 +73,8 @@ func init() {
 
 	cwd, err := os.Getwd()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "couldn't get current working directory: %s\n", err)
+		fmt.Fprintf(os.Stderr,
+			"can't get current working directory: %s\n", err)
 		os.Exit(1)
 	}
 	workdir = cwd
@@ -91,8 +92,8 @@ func main() {
 
 		dependencies, _ = args["<dependency>"].([]string)
 
-		recursive        = args["--recursive"].(bool)
-		testDependencies = args["--testing"].(bool)
+		recursive       = args["--recursive"].(bool)
+		includeTestDeps = args["--testing"].(bool)
 	)
 
 	verbose = args["--verbose"].(bool)
@@ -100,20 +101,20 @@ func main() {
 	var err error
 	switch {
 	case modeInstall:
-		err = handleInstall(recursive, testDependencies, dependencies)
+		err = handleInstall(recursive, includeTestDeps, dependencies)
 
 	case modeUpdate:
-		err = handleUpdate(recursive, testDependencies, dependencies)
+		err = handleUpdate(recursive, includeTestDeps, dependencies)
 
 	case modeQuery:
 		onlyVendored := args["-o"].(bool)
-		err = handleQuery(recursive, testDependencies, onlyVendored)
+		err = handleQuery(recursive, includeTestDeps, onlyVendored)
 
 	case modeRemove:
 		err = handleRemove(dependencies)
 
 	case modeClean:
-		err = handleClean(recursive, testDependencies)
+		err = handleClean(recursive, includeTestDeps)
 	}
 
 	if err != nil {
@@ -121,8 +122,9 @@ func main() {
 	}
 }
 
-func handleInstall(recursive bool, testDependencies bool, dependencies []string) error {
-	imports, err := parseImports(recursive, testDependencies)
+func handleInstall(recursive bool, includeTestDeps bool,
+	dependencies []string) error {
+	imports, err := parseImports(recursive, includeTestDeps)
 	if err != nil {
 		return err
 	}
@@ -190,10 +192,11 @@ func handleInstall(recursive bool, testDependencies bool, dependencies []string)
 	return nil
 }
 
-func handleUpdate(recursive bool, testDependencies bool, dependencies []string) error {
+func handleUpdate(recursive bool, includeTestDeps bool,
+	dependencies []string) error {
 	var err error
 	if len(dependencies) == 0 {
-		dependencies, err = parseImports(recursive, testDependencies)
+		dependencies, err = parseImports(recursive, includeTestDeps)
 		if err != nil {
 			return err
 		}
@@ -261,7 +264,7 @@ func handleRemove(dependencies []string) error {
 	return nil
 }
 
-func handleQuery(recursive, testDependencies, onlyVendored bool) error {
+func handleQuery(recursive, includeTestDeps, onlyVendored bool) error {
 	submodules, err := getVendorSubmodules()
 	if err != nil {
 		return err
@@ -275,7 +278,7 @@ func handleQuery(recursive, testDependencies, onlyVendored bool) error {
 			fmt.Printf(format, submodule, commit)
 		}
 	} else {
-		imports, err := parseImports(recursive, testDependencies)
+		imports, err := parseImports(recursive, includeTestDeps)
 		if err != nil {
 			return err
 		}
@@ -303,8 +306,8 @@ func handleQuery(recursive, testDependencies, onlyVendored bool) error {
 	return nil
 }
 
-func handleClean(recursive, testDependencies bool) error {
-	imports, err := parseImports(recursive, testDependencies)
+func handleClean(recursive, includeTestDeps bool) error {
+	imports, err := parseImports(recursive, includeTestDeps)
 	if err != nil {
 		return err
 	}

--- a/main.go
+++ b/main.go
@@ -40,6 +40,7 @@ Options:
     -Q --query      List all dependencies.
         -o          List only already-vendored dependencies.
     -C --clean      Detect all unused vendored dependencies and remove it.
+	-t --testing    Include testing dependencies.
     -r --recursive  Be recursive.
     -h --help       Show help message.
     -v --verbose    Be verbose.
@@ -82,7 +83,8 @@ func main() {
 
 		dependencies, _ = args["<dependency>"].([]string)
 
-		recursive = args["--recursive"].(bool)
+		recursive        = args["--recursive"].(bool)
+		testDependencies = args["--testing"].(bool)
 	)
 
 	verbose = args["--verbose"].(bool)
@@ -90,20 +92,20 @@ func main() {
 	var err error
 	switch {
 	case modeInstall:
-		err = handleInstall(recursive, dependencies)
+		err = handleInstall(recursive, testDependencies, dependencies)
 
 	case modeUpdate:
-		err = handleUpdate(recursive, dependencies)
+		err = handleUpdate(recursive, testDependencies, dependencies)
 
 	case modeQuery:
 		onlyVendored := args["-o"].(bool)
-		err = handleQuery(recursive, onlyVendored)
+		err = handleQuery(recursive, testDependencies, onlyVendored)
 
 	case modeRemove:
 		err = handleRemove(dependencies)
 
 	case modeClean:
-		err = handleClean(recursive)
+		err = handleClean(recursive, testDependencies)
 	}
 
 	if err != nil {
@@ -111,8 +113,8 @@ func main() {
 	}
 }
 
-func handleInstall(recursive bool, dependencies []string) error {
-	imports, err := parseImports(recursive)
+func handleInstall(recursive bool, testDependencies bool, dependencies []string) error {
+	imports, err := parseImports(recursive, testDependencies)
 	if err != nil {
 		return err
 	}
@@ -180,10 +182,10 @@ func handleInstall(recursive bool, dependencies []string) error {
 	return nil
 }
 
-func handleUpdate(recursive bool, dependencies []string) error {
+func handleUpdate(recursive bool, testDependencies bool, dependencies []string) error {
 	var err error
 	if len(dependencies) == 0 {
-		dependencies, err = parseImports(recursive)
+		dependencies, err = parseImports(recursive, testDependencies)
 		if err != nil {
 			return err
 		}
@@ -251,7 +253,7 @@ func handleRemove(dependencies []string) error {
 	return nil
 }
 
-func handleQuery(recursive, onlyVendored bool) error {
+func handleQuery(recursive, testDependencies, onlyVendored bool) error {
 	submodules, err := getVendorSubmodules()
 	if err != nil {
 		return err
@@ -265,7 +267,7 @@ func handleQuery(recursive, onlyVendored bool) error {
 			fmt.Printf(format, submodule, commit)
 		}
 	} else {
-		imports, err := parseImports(recursive)
+		imports, err := parseImports(recursive, testDependencies)
 		if err != nil {
 			return err
 		}
@@ -293,8 +295,8 @@ func handleQuery(recursive, onlyVendored bool) error {
 	return nil
 }
 
-func handleClean(recursive bool) error {
-	imports, err := parseImports(recursive)
+func handleClean(recursive, testDependencies bool) error {
+	imports, err := parseImports(recursive, testDependencies)
 	if err != nil {
 		return err
 	}

--- a/main.go
+++ b/main.go
@@ -51,6 +51,7 @@ Options:
 var (
 	verbose bool
 	inTests bool
+	workdir string
 )
 
 func init() {
@@ -69,6 +70,12 @@ func init() {
 		}
 	}
 	os.Args = newArgs
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(fmt.Errorf("couldn't get current working directory: %s", err))
+	}
+	workdir = cwd
 }
 
 func main() {

--- a/main.go
+++ b/main.go
@@ -40,7 +40,7 @@ Options:
     -Q --query      List all dependencies.
         -o          List only already-vendored dependencies.
     -C --clean      Detect all unused vendored dependencies and remove it.
-	-t --testing    Include testing dependencies.
+    -t --testing    Include testing dependencies.
     -r --recursive  Be recursive.
     -h --help       Show help message.
     -v --verbose    Be verbose.

--- a/main.go
+++ b/main.go
@@ -40,7 +40,7 @@ Options:
     -Q --query      List all dependencies.
         -o          List only already-vendored dependencies.
     -C --clean      Detect all unused vendored dependencies and remove it.
-    -t --testing    Include testing dependencies.
+    -t --testing    Include dependencies from tests.
     -r --recursive  Be recursive.
     -h --help       Show help message.
     -v --verbose    Be verbose.

--- a/main.go
+++ b/main.go
@@ -73,7 +73,8 @@ func init() {
 
 	cwd, err := os.Getwd()
 	if err != nil {
-		panic(fmt.Errorf("couldn't get current working directory: %s", err))
+		fmt.Fprintf(os.Stderr, "couldn't get current working directory: %s\n", err)
+		os.Exit(1)
 	}
 	workdir = cwd
 }

--- a/submodule.go
+++ b/submodule.go
@@ -4,7 +4,6 @@ import (
 	"errors"
 	"fmt"
 	"go/build"
-	"os"
 	"os/exec"
 	"path/filepath"
 	"strings"
@@ -171,15 +170,10 @@ func removeVendorSubmodule(importpath string) error {
 }
 
 func updateVendorSubmodule(importpath string) error {
-	cwd, err := os.Getwd()
-	if err != nil {
-		return err
-	}
-
 	cmd := exec.Command("git", "pull", "origin", "master")
-	cmd.Dir = filepath.Join(cwd, "vendor", importpath)
+	cmd.Dir = filepath.Join(workdir, "vendor", importpath)
 
-	_, err = execute(cmd)
+	_, err := execute(cmd)
 
 	return err
 }
@@ -190,8 +184,7 @@ func getRootImportpath(importpath string) (string, error) {
 		return "", err
 	}
 
-	cwd, _ := os.Getwd()
-	vendorPath := strings.TrimPrefix(cwd, pkg.SrcRoot+"/") + "/vendor/"
+	vendorPath := strings.TrimPrefix(workdir, pkg.SrcRoot+"/") + "/vendor/"
 
 	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
 	cmd.Dir = filepath.Join(pkg.SrcRoot, importpath)

--- a/submodule.go
+++ b/submodule.go
@@ -190,6 +190,9 @@ func getRootImportpath(importpath string) (string, error) {
 		return "", err
 	}
 
+	cwd, _ := os.Getwd()
+	vendorPath := strings.TrimPrefix(cwd, pkg.SrcRoot+"/") + "/vendor/"
+
 	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
 	cmd.Dir = filepath.Join(pkg.SrcRoot, importpath)
 
@@ -198,8 +201,7 @@ func getRootImportpath(importpath string) (string, error) {
 		return "", err
 	}
 
-	return strings.Trim(
+	return strings.TrimPrefix(strings.Trim(
 		strings.TrimSpace(strings.TrimPrefix(rootdir, pkg.SrcRoot)),
-		"/",
-	), nil
+		"/"), vendorPath), nil
 }

--- a/utils.go
+++ b/utils.go
@@ -16,8 +16,8 @@ func execute(cmd *exec.Cmd) (string, error) {
 		}
 	}
 
-	stdout, _, err := executil.Run(cmd)
-	return string(stdout), err
+	stdout, stderr, err := executil.Run(cmd)
+	return string(stdout) + string(stderr), err
 }
 
 func getMaxLength(elements []string) int {

